### PR TITLE
[CSLD-30] Rewrite section on soft fork updates

### PR DESCRIPTION
**Pull request by Pastafarianist**

### CODE REVIEW

File: _docs/cardano/2017-01-06-updating.md [hunk 1]
```diff
@@ -110,28 +110,25 @@ new version blocks are still compatible with old version clients.
 A hard fork is one that doesn’t maintain backward compatibility with
 the previous version.
 
-BIP-99 provides an excellent criteria to distinguish between these two
+BIP-99 provides excellent criteria to distinguish between these two
 types of fork:
 
-### Soft fork
+* A **soft fork** is a consensus fork wherein everything that was previously
+  invalid remains invalid while blocks that would have been previously considered
```
Comment (rockbmb) [inline _docs/cardano/2017-01-06-updating.md]: Be careful with your quantification - not all blocks that were previously valid become invalid. Only "some blocks", although you don't need to go on to explain why. Maybe just explain that a soft-fork introuces new rules that cause this.

File: _docs/cardano/2017-01-06-updating.md [hunk 2]
```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
...
+* **Adopted**, when the softfork resolution rule (see below) for a confirmed
+  block version is triggered.
+* **Confirmed**, when there is an update proposal that contains a confirmed
+  version of the software and this block version. Note that "confirmed version of
+  the software" is a technical term defined elsewhere. If there are multiple block
+  versions such that corresponding software is confirmed, but these versions
+  aren't adopted, we call them **competing**. For instance, there might be
+  versions `2.0.0`, `2.0.1`, `1.2.0`, `1.2.1`, `1.1.1` and `1.1.2`, with the last
+  adopted version being `1.1.3`. In this case the versions that are competing are
+  `2.0.0`, `2.0.1`, `1.2.0` and `1.2.1`. `1.1.1` and `1.1.2` aren't competing,
```
Comment (rockbmb) [inline _docs/cardano/2017-01-06-updating.md]: It's unclear where your enumeration of competing versions starts and where it ends.

File: _docs/cardano/2017-01-06-updating.md (hunk 1 above)
Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: 👍

Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Fixed.

File: _docs/cardano/2017-01-06-updating.md (hunk 2 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Fixed.

File: _docs/cardano/2017-01-06-updating.md [hunk 3]
```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
...
+
+And then imagine somebody proposing a transaction to this address, possibly
+doing that with an intention to secure funds from being spent until some
+conditions are met — and then watching them being spent in some other
+transaction in a block with version 1. This is the point. We cannot make use of
+a restriction without waiting for the network to start assuming the old version
+to be deprecated (since we can only start rejecting blocks when their version is
+deprecated).
+
+Another problem is that in our implementations all blocks have a field called
```
Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: Can it be reformulated somehow? Reader can think this _another problem_ is about the same concern, i. e. that we need to deprecate old version. But this _another problem_ is about different thing. It's about that we need to prohibit usage of new features from new version until it's adopted.

Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: AFAIU

File: _docs/cardano/2017-01-06-updating.md [hunk 4]
```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
...
- 1. Once the update is approved, the protocol version (say v0.5) can be used.
- 2. Before a soft fork is resolved (i.e. the resolution rule hit), the
-   updated nodes should issue blocks with the new version v5, but treat
-   and validate blocks of version v5 as blocks of version v4.
- 3. Non-updated updated nodes accept new blocks:
-   a. Before the soft fork is resolved, discarding as per v4.
-   b. After the soft fork is resolved, as per v5 (i.e. allowing a block to
-  have unknown attributes, and so on).
+
+1. Once the update is approved, the protocol version (say `0.5.0`) can be used.
```
Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: No, only when update is *confirmed*, he-he. 😜

File: _docs/cardano/2017-01-06-updating.md [hunk 5]
```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
...
+2. Before a soft fork is resolved (i.e. the resolution rule hit), the updated
+   nodes should issue blocks with the new version `0.5.0`, but treat and validate
+   blocks of version `0.5.0` as blocks of version `0.4.0`.
+3. Non-updated updated nodes accept new blocks:
```
Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: > Non-updated updated

wut

File: _docs/cardano/2017-01-06-updating.md [hunk 6]
```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
...
+    1. Before the soft fork is resolved, discarding as per `0.4.0`.
```
Comment (gromakovsky) [inline _docs/cardano/2017-01-06-updating.md]: I guess something is missing in this sentence…

File: _docs/cardano/2017-01-06-updating.md (hunk 4 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: That's a part from the previous version that I haven't edited yet, except for fixing the version numbers.

File: _docs/cardano/2017-01-06-updating.md (hunk 3 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Fixed.

File: _docs/cardano/2017-01-06-updating.md (hunk 4 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Fixed.

File: _docs/cardano/2017-01-06-updating.md (hunk 5 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Rewritten that section entirely.

File: _docs/cardano/2017-01-06-updating.md (hunk 6 above)
Comment (Pastafarianist) [inline _docs/cardano/2017-01-06-updating.md]: Rewritten that section entirely.

**Pull request merged**